bika/lims/browser/js/bika.lims.analysisrequest.add.js

```diff
@@ -80,40 +82,71 @@
-      $("tr[fieldname=Profiles] img.deletebtn").live("click", this.on_analysis_profile_removed);
-      $("img.copybutton").on("click", this.on_copy_button_click);
+      $("body").on("click", ".service-listing-header", this.on_service_listing_header_click);
+      $("body").on("click", "tr.category", this.on_service_category_click);
+      $("body").on("click", "[name='save_button']", this.on_form_submit);
+      $("body").on("click", "tr[fieldname=AdHoc] input[type='checkbox']", this.recalculate_records);
+      $("body").on("click", "tr[fieldname=Composite] input[type='checkbox']", this.recalculate_records);
+      $("body").on("click", "tr[fieldname=InvoiceExclude] input[type='checkbox']", this.recalculate_records);
+      $("body").on("click", "tr[fieldname=Analyses] input[type='checkbox']", this.on_analysis_checkbox_click);
+      $("body").on("selected change", "tr[fieldname=Client] input[type='text']", this.on_client_changed);
+      $("body").on("selected change", "tr[fieldname=Contact] input[type='text']", this.on_contact_changed);
+      $("body").on("click", "tr[fieldname=ReportDryMatter] input[type='checkbox']", this.on_reportdrymatter_click);
+      $("body").on("change", "input.min", this.on_analysis_specification_changed);
+      $("body").on("change", "input.max", this.on_analysis_specification_changed);
+      $("body").on("change", "input.err", this.on_analysis_specification_changed);
+      $("body").on("click", ".service-lockbtn", this.on_analysis_lock_button_click);
+      $("body").on("click", ".service-infobtn", this.on_analysis_details_click);
+      $("body").on("selected change", "tr[fieldname=Sample] input[type='text']", this.on_sample_changed);
+      $("body").on("selected change", "tr[fieldname=SampleType] input[type='text']", this.on_sampletype_changed);
+      $("body").on("selected change", "tr[fieldname=Specification] input[type='text']", this.on_specification_changed);
+      $("body").on("selected change", "tr[fieldname=Template] input[type='text']", this.on_analysis_template_changed);
+      $("body").on("selected", "tr[fieldname=Profiles] input[type='text']", this.on_analysis_profile_selected);
+      $("body").on("click", "tr[fieldname=Profiles] img.deletebtn", this.on_analysis_profile_removed);
+      $("body").on("click", "img.copybutton", this.on_copy_button_click);
 
       /* internal events */
-      $(this).on("form:changed", this.recalculate_records);
-      $(this).on("data:updated", this.update_form);
-      $(this).on("data:updated", this.recalculate_prices);
-      $(this).on("data:updated", this.hide_all_service_info);
+      $(this).on("form:changed", this.debounce(this.recalculate_records, 500));
+      $(this).on("data:updated", this.debounce(this.recalculate_prices, 3000));
+      $(this).on("data:updated", this.debounce(this.update_form, 300));
+      $(this).on("data:updated", this.debounce(this.hide_all_service_info, 300));
```

Comment:
If I am not wrong, "300" in this case are the ms. the application will wait between successive `data:updated` calls before calling `hide_all_services_info`. So, if `data_updated` is called multiple times within those 300 ms, `hide_all_service_info` will only be called once, 300ms after the last call, when the elapsed time between that last  `data:updated` call and the next is greater than 300ms). Is my reasoning right?

Reply:
Exactly. It is called _after_ the given time period. We could also pass in `true` as the third parameter to execute it immediately and disallow any further calls in the timeframe